cpp/src/io/csv/csv_gpu.h

```diff
@@ -23,6 +23,112 @@ namespace cudf {
 namespace io {
 namespace csv {
 namespace gpu {
+
+/**
+ * CSV row parsing context
+ * NONE: No special context (normal parsing)
+ * QUOTE: Within a quoted field
+ * COMMENT: Within a comment line (discard every character until terminator)
+ * EOF: End state (EOF reached)
+ **/
+enum { ROW_CTX_NONE = 0, ROW_CTX_QUOTE = 1, ROW_CTX_COMMENT = 2, ROW_CTX_EOF = 3 };
+
+constexpr uint32_t rowofs_block_dim = 512;
+/// Character block size for gather_row_offsets
+constexpr uint32_t rowofs_block_bytes = rowofs_block_dim * 32;  // 16KB/threadblock
+
+/**
+ * @brief return a row context from a {count, id} pair
+ *
+ * The 32-bit row context consists of the 2-bit parser state stored in the lower 2-bits
+ * and a 30-bit row count in the upper 30 bits.
+ *
+ **/
+inline __host__ __device__ uint32_t make_row_context(uint32_t row_count, uint32_t out_ctx)
+{
+  return row_count * 4 + out_ctx;
+}
+
+/**
+ * @brief pack multiple row contexts together
+ *
+ * The 64-bit packed row context format represents the four possible output row context state
+ * from each of the four possible input row context state.
+ * Each output state consists of the 2-bit row context state along with a 18-bit row count
+ * value (row count is assumed to be a local count that fits in 18-bit)
+ * The four 20-bit values are concatenated to form a 80-bit value, truncated to 64-bit
+ * since a block starting in a EOF state can only have a zero row count (and the output
+ * state corresponding to an EOF input state can only be EOF, so only the first 3 output
+ * states are included as parameters, and the EOF->EOF state transition is hardcoded)
+ *
+ **/
+inline __host__ __device__ uint64_t pack_row_contexts(uint32_t ctx0, uint32_t ctx1, uint32_t ctx2)
+{
+  return (ctx0) | (static_cast<uint64_t>(ctx1) << 20) | (static_cast<uint64_t>(ctx2) << 40) |
+         (static_cast<uint64_t>(ROW_CTX_EOF) << 60);
+}
+
+/**
+ * @brief Unpack a row context  (select one of the 4 contexts in packed form)
+ **/
+inline __host__ __device__ uint32_t get_row_context(uint64_t packed_ctx, uint32_t ctxid)
+{
+  return static_cast<uint32_t>((packed_ctx >> (ctxid * 20)) & ((1 << 20) - 1));
+}
+
+/**
+ * @brief Select a row context given another input context (updating 62-bit row count)
+ **/
+inline __host__ __device__ uint64_t select_row_context(uint64_t sel_ctx, uint64_t packed_ctx)
```

Comment:
I cannot figure out what this function does. Can you please explain in more detail in the comment above?

Reply:
Added a more descriptive comment